src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -371,16 +368,53 @@ private async ValueTask<EvaluationEvent> EvaluateFeature<TContext>(string featur
                     await sessionManager.SetAsync(evaluationEvent.FeatureDefinition.Name, evaluationEvent.Enabled).ConfigureAwait(false);
                 }
 
-                if (evaluationEvent.FeatureDefinition.Telemetry != null &&
-                    evaluationEvent.FeatureDefinition.Telemetry.Enabled)
+                // Only add an activity event if telemetry is enabled for the feature
+                if (telemetryEnabled)
                 {
-                    PublishTelemetry(evaluationEvent, cancellationToken);
+                    AddEvaluationActivityEvent(evaluationEvent);
                 }
             }
 
             return evaluationEvent;
         }
 
+        private void AddEvaluationActivityEvent(EvaluationEvent evaluationEvent)
+        {
+            // Break if there is no activity, this shouldn't happen
+            if (Activity.Current == null)
+            {
+                return;
+            }
+
+            ActivityTagsCollection tags = new ActivityTagsCollection
+            {
+                { "TargetingId", evaluationEvent.TargetingContext?.UserId },
+                { "FeatureName", evaluationEvent.FeatureDefinition.Name },
+                { "Variant", evaluationEvent.Variant?.Name },
+                { "Enabled", evaluationEvent.Enabled },
+                { "VariantAssignmentReason", evaluationEvent.VariantAssignmentReason },
+                { "Version", ActivitySource.Version }
+            };
+
+            if (evaluationEvent.FeatureDefinition.Telemetry.Metadata != null)
+            {
+                foreach (KeyValuePair<string, string> kvp in evaluationEvent.FeatureDefinition.Telemetry.Metadata)
+                {
+                    try
+                    {
+                        tags.Add(kvp.Key, kvp.Value);
```

Comment:
I'd recommend using 'ContainsKey` explicitly instead of relying on the exception from `Add` to log the warning, so we avoid relying on exceptions as control flow even if it is slightly more verbose. Relying on the exception is also substantially slower and could become a problem in critical paths if it keeps throwing for several features.

Reply:
Good point- updated.